xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/store/XWikiVersioningStoreInterface.java

```diff
@@ -47,6 +55,74 @@ void saveXWikiDocArchive(XWikiDocumentArchive archivedoc, boolean bTransaction,
 
     Version[] getXWikiDocVersions(XWikiDocument doc, XWikiContext context) throws XWikiException;
 
+    private Collection<String> getXWikiDocStringVersions(XWikiDocument doc, RevisionCriteria criteria,
+        XWikiContext context) throws XWikiException
+    {
+        List<String> results = new ArrayList<>();
+
+        Version[] revisions = getXWikiDocVersions(doc, context);
+        XWikiDocumentArchive archive = getXWikiDocumentArchive(doc, context);
+
+        XWikiRCSNodeInfo nextNodeinfo = null;
+        XWikiRCSNodeInfo nodeinfo;
+        for (Version revision : revisions) {
+            nodeinfo = nextNodeinfo;
+            nextNodeinfo = archive.getNode(revision);
+
+            if (nodeinfo != null) {
+                // Minor/Major version matching
+                if (criteria.getIncludeMinorVersions() || nextNodeinfo == null || !nextNodeinfo.isMinorEdit()) {
+                    // Author matching
+                    if (criteria.getAuthor().isEmpty() || criteria.getAuthor().equals(nodeinfo.getAuthor())) {
+                        // Date range matching
+                        Date versionDate = nodeinfo.getDate();
+                        if (versionDate.after(criteria.getMinDate()) && versionDate.before(criteria.getMaxDate())) {
+                            results.add(nodeinfo.getVersion().toString());
+                        }
+                    }
+                }
+            }
+        }
+
+        return criteria.getRange().subList(results);
+    }
+
+    /**
+     * Gets versions from a given document matching criteria like author, minimum creation date, etc.
+     *
+     * @param doc the document
+     * @param criteria criteria used to match versions
+     * @param context the XWiki context
+     * @return the matching versions
+     * @since 15.10.7
+     * @since 16.2.0RC1
+     */
+    @Unstable
+    default Collection<Version> getXWikiDocVersions(XWikiDocument doc, RevisionCriteria criteria, XWikiContext context)
+        throws XWikiException
+    {
+        return getXWikiDocStringVersions(doc, criteria, context).stream()
+            .map(Version::new)
+            .toList();
```

Comment:
Be careful as this is not going to work in 15.10.x (Java 11), I generally try to use the same code for both branches for code which is intended to be cherry-picked (make future bugfixing easier).